src/rust/engine/fs/store/src/lib.rs

```diff
@@ -148,35 +149,78 @@ pub struct UploadSummary {
   pub upload_wall_time: Duration,
 }
 
+///
+/// Wraps a `remote::ByteStore` with state to help avoid uploading common blobs multiple times.
+///
+/// If a blob is generated that many downstream actions depend on, we would otherwise get an
+/// expanded set of digests from each of those actions that includes the new blob. If those actions
+/// all execute in a time window smaller than the time taken to upload the blob, the effort would be
+/// duplicated leading to both wasted resources locally buffering up the blob as well as wasted
+/// effort on the remote server depending on its handling of this.
+///
 #[derive(Clone, Debug)]
 struct RemoteStore {
   store: remote::ByteStore,
-  in_flight_uploads: Arc<parking_lot::Mutex<HashSet<Digest>>>,
+  in_flight_uploads: Arc<Mutex<HashMap<Digest, Weak<OnceCell<()>>>>>,
+  in_flight_downloads: Arc<Mutex<HashMap<Digest, Weak<OnceCell<()>>>>>,
 }
 
 impl RemoteStore {
   fn new(store: remote::ByteStore) -> Self {
     Self {
       store,
-      in_flight_uploads: Arc::new(parking_lot::Mutex::new(HashSet::new())),
+      in_flight_uploads: Arc::default(),
+      in_flight_downloads: Arc::default(),
     }
   }
 
-  fn reserve_uploads(&self, candidates: HashSet<Digest>) -> HashSet<Digest> {
-    let mut active_uploads = self.in_flight_uploads.lock();
-    let to_upload = candidates
-      .difference(&active_uploads)
-      .cloned()
-      .collect::<HashSet<_>>();
-    active_uploads.extend(&to_upload);
-    to_upload
+  ///
+  /// Returns a strongly held cell from a map of weakly held cells, creating it if necessary.
+  ///
+  fn cell_from(
+    cells: &Mutex<HashMap<Digest, Weak<OnceCell<()>>>>,
+    digest: Digest,
+  ) -> Arc<OnceCell<()>> {
+    let mut cells = cells.lock();
+    if let Some(cell) = cells.get(&digest).and_then(|weak_cell| weak_cell.upgrade()) {
+      cell
+    } else {
+      let cell = Arc::new(OnceCell::new());
+      cells.insert(digest, Arc::downgrade(&cell));
+      cell
+    }
   }
 
-  fn release_uploads(&self, uploads: HashSet<Digest>) {
-    self
-      .in_flight_uploads
-      .lock()
-      .retain(|d| !uploads.contains(d));
+  ///
+  /// Guards an attempt to upload the given `Digest`, skipping the upload if another attempt has
+  /// been successful. Will not return until either an attempt has succeed, or this attempt has
+  /// failed.
+  ///
+  async fn maybe_upload<E>(
+    &self,
+    digest: Digest,
+    upload: impl Future<Output = Result<(), E>>,
+  ) -> Result<(), E> {
+    Self::cell_from(&self.in_flight_uploads, digest)
+      .get_or_try_init(upload)
+      .await
+      .map(|&()| ())
```

Comment:
Yes. But when discarding return values like this, I prefer to be explicit. If the inner value were to be refactored to return a `Result<..>`, the `.map(|_| ())` would continue to compile, while discarding something meaningful (because it's a bit like `let _ = ..`, which is how you'd silence a "must use" warning).